src/helpers/string-to-number-validator.ts

```diff
@@ -1,14 +1,37 @@
-import { Type } from "class-transformer";
-import { IsNumber } from "class-validator";
+import { Transform, Type } from "class-transformer";
+import { isNumber, registerDecorator, ValidationOptions } from "class-validator";
 
 /**
  * Checks if a value can be converted to a number
  */
-export const StringToNumber = (): PropertyDecorator => {
-    return (propertyValue: unknown, propertyName: string): void => {
-        // Cast the value to a number
-        Type(() => Number)(propertyValue, propertyName);
-        // Validate whether the value is a number
-        IsNumber()(propertyValue, propertyName);
+export function StringToNumber(
+    validationOptions?: ValidationOptions & {
+        allowNulls: boolean;
+    }
+) {
+    return function (object: unknown, propertyName: string): void {
+        registerDecorator({
+            name: "stringToNumber",
+            target: object.constructor,
+            propertyName: propertyName,
+            constraints: [],
+            options: validationOptions,
+            validator: {
+                validate(value: unknown) {
+                    if (
+                        validationOptions?.allowNulls &&
+                        (value === null || value === "null")
+                    ) {
+                        // Force value to null. This mutates the value which exists only inside class-transformer.
+                        Transform(() => null)(object, propertyName);
+                        return true;
+                    }
+
+                    // Cast the value to a number, mutating it for the next operations, and validate it
+                    Type(() => Number)(object, propertyName);
+                    return isNumber(Number(value));
```

Comment:
Can this throw exceptions if value is not valid? Or will it just return NaN?
If exceptions are possible, add try-catch , possibly setting the value to 0.

Reply:
If an invalid string number is passed to `Number()`, the output is guaranteed to be `NaN`. No exceptions are thrown.
As mentioned in my reply to your other comment, I have reverted the validator.